src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,288 @@
+        targetBuffer->length = 0;
+        targetBuffer->data = nullptr;
+        return majorStatus;
+    }
+
+    targetBuffer->length = gssBuffer->length;
+    targetBuffer->data = static_cast<uint8_t*>(gssBuffer->value);
+    return majorStatus;
+}
+
+static uint32_t NetSecurityNative_AcquireCredSpNego(uint32_t* minorStatus,
+                                                    GssName* desiredName,
+                                                    gss_cred_usage_t credUsage,
+                                                    GssCredId** outputCredHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(desiredName != nullptr);
+    assert(outputCredHandle != nullptr);
+#if HAVE_GSS_SPNEGO_MECHANISM
+    gss_OID_set_desc gss_mech_spnego_OID_set_desc = {.count = 1, .elements = GSS_SPNEGO_MECHANISM};
+#else
+    gss_OID_desc gss_mech_spnego_OID_desc = {.length = 6, .elements = static_cast<void*>(gss_mech_value)};
+    gss_OID_set_desc gss_mech_spnego_OID_set_desc = {.count = 1, .elements = &gss_mech_spnego_OID_desc};
+#endif
+    return gss_acquire_cred(
+        minorStatus, desiredName, 0, &gss_mech_spnego_OID_set_desc, credUsage, outputCredHandle, nullptr, nullptr);
+}
+
+extern "C" uint32_t NetSecurityNative_InitiateCredSpNego(uint32_t* minorStatus,
+                                                         GssName*  desiredName,
+                                                         GssCredId** outputCredHandle)
+{
+    return NetSecurityNative_AcquireCredSpNego(minorStatus, desiredName, GSS_C_INITIATE, outputCredHandle);
+}
+
+extern "C" uint32_t NetSecurityNative_DeleteSecContext(uint32_t* minorStatus, GssCtxId** contextHandle)
+{
+    assert(minorStatus != nullptr);
+    assert(contextHandle != nullptr);
+    return gss_delete_sec_context(minorStatus, contextHandle, GSS_C_NO_BUFFER);
+}
+
+extern "C" uint32_t NetSecurityNative_DisplayStatus(uint32_t* minorStatus,
+                                                    uint32_t statusValue,
+                                                    int32_t isGssMechCode,
+                                                    struct PAL_GssBuffer* outBuffer)
+{
+    assert(minorStatus != nullptr);
+    assert(isGssMechCode == 0 || isGssMechCode == 1);
+    assert(outBuffer != nullptr);
+
+    int statusType = isGssMechCode ? GSS_C_MECH_CODE : GSS_C_GSS_CODE;
+    GssBuffer gssBuffer {.length = 0, .value = nullptr};
+    uint32_t majorStatus =
+        gss_display_status(minorStatus, statusValue, statusType, GSS_C_NO_OID, nullptr, &gssBuffer);
+    return NetSecurityNative_HandleError(majorStatus, &gssBuffer, outBuffer);
+}
+
+extern "C" uint32_t
+NetSecurityNative_ImportUserName(uint32_t* minorStatus, char* inputName, uint32_t inputNameLen, GssName** outputName)
+{
+    assert(minorStatus != nullptr);
+    assert(outputName != nullptr);
+    assert(inputName != nullptr);
+    assert(outputName != nullptr);
+    GssBuffer inputNameBuffer{.length = inputNameLen, .value = inputName};
```

Comment:
Is there anything we should assert about `*outputName`?  Should it be null? Not null? Doesn't matter?